### PR TITLE
Adds support for looking up past Blocks in expando objects

### LINKED ISSUE #31972: Function with properties type inferred only at top level
**TypeScript Version:**  3.5.1

**Search Terms:**

- 2739
- function properties block
- top level function

**Code**

```ts
interface Person {
  first: string;
  last: string;
}

const dice = () => Math.floor(Math.random() * 6);
dice.first = 'Rando';
dice.last = 'Calrissian';
const diceP: Person = dice;  // OK

// but inside a block...
{
  const dice = () => Math.floor(Math.random() * 6);
  dice.first = 'Rando';  // Property 'first' does not exist on type '() => number'. 
  dice.last = 'Calrissian';  // Property 'last' does not exist on type '() => number'.
  const diceP: Person = dice;  // Type '() => number' is missing the following properties from type 'Person': first, last
}
```

**Expected behavior:**

I would expect the same behavior both inside and outside the block.

**Actual behavior:**

Assigning properties to a function at the top level is allowed, but not inside a block. Frankly, I'm impressed that the first example works at all. But I don't see why they would be different!

**Playground Link:** [link][1]

**Related Issues:** 

- https://github.com/microsoft/TypeScript/issues/31778

[1]: http://www.typescriptlang.org/play/#src=interface%20Person%20%7B%0D%0A%20%20first%3A%20string%3B%0D%0A%20%20last%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Aconst%20dice%20%3D%20()%20%3D%3E%20Math.floor(Math.random()%20*%206)%3B%0D%0Adice.first%20%3D%20'Rando'%3B%0D%0Adice.last%20%3D%20'Calrissian'%3B%0D%0Aconst%20diceP%3A%20Person%20%3D%20dice%3B%20%20%2F%2F%20OK%0D%0A%0D%0A%2F%2F%20but%20inside%20a%20block...%0D%0A%7B%0D%0A%20%20const%20dice%20%3D%20()%20%3D%3E%20Math.floor(Math.random()%20*%206)%3B%0D%0A%20%20dice.first%20%3D%20'Rando'%3B%20%20%2F%2F%20Property%20'first'%20does%20not%20exist%20on%20type%20'()%20%3D%3E%20number'.%20%0D%0A%20%20dice.last%20%3D%20'Calrissian'%3B%20%20%2F%2F%20Property%20'last'%20does%20not%20exist%20on%20type%20'()%20%3D%3E%20number'.%0D%0A%20%20const%20diceP%3A%20Person%20%3D%20dice%3B%20%20%2F%2F%20Type%20'()%20%3D%3E%20number'%20is%20missing%20the%20following%20properties%20from%20type%20'Person'%3A%20first%2C%20last%0D%0A%7D%0D%0A

### PR DESCRIPTION
**Pull request by orta**
Fixes #31972 

The `Block` in the AST stops the the lookup for the symbol in the containing scope for the expando function, this PR makes the symbol lookup fallback to checking the blockscope scope when needed as well

/cc @andrewbranch

### CODE REVIEW

**Comment by sandersn** [comment]
Doesn't this only fail to work in TS? Can you add a test for JS as well and see whether it fails without your fix?

**Comment by orta** [comment]
Confirmed this fails in JS in the current playground, and double confirmed that it fixes it on that side too

```
// Creates a type { first:string, last: string }
/**
 * @typedef {Object} Human - creates a new type named 'SpecialType'
 * @property {string} first - a string property of SpecialType
 * @property...
```

[Playground Link](https://www.typescriptlang.org/v2/play?ssl=1&ssc=1&pln=27&pc=1&useJavaScript=true#code/PTAEGECcFMEMBdoGdS1PAngB2qA3qAGYCWkS8AXOZMQHYDmANKADazkWjV32gC+AWABQwAFSjhoUaAACmHABNohfAHkARgCtoAY3h9QACQCuAW1i1QAWlA6YCZKlC1oAd3TZctWKegLQAOQAyjg6xLAsACqeAZLSMliQAPY4kJj43AwGJGTw1k6ZvIkp0GkYoEkqIbrhUZ5xssWp6XiFBmzk+Wi0ZuqloE2l6ZWg1WER0ThxwMLCYhJCUo2wkD74Jua0Blgra8ZIfqgoaABuEcQKCMRJlvBJSSzTwoTGtHrXlgpJGxaRABY8JAACh2q1MAEp8IIhMIdDdOgpiDpcABeUBAyEogB8oAAsgg-gA6Qgse6QIH4+BE1a0L6mDFLABs4IA3MJEcjiaROmiAgAlCxfAJsoQc6CEjp5XngCI0JBIcK0YWzUXfMy-AEMYFi8EqkCgdTGPJ0BVKJzqUk6ADWhNtoAwSWMtgsATyOgiLFAxCNTj+6towjwkls8LyYtAaIZ2LxBOJpKS5Mp1MFSXpkOkzJFoFAYq5uQjgQFtKSysWOaR4slBYCMpYcoVLpZoGDwa+P1o-0BQJ1wmhQA)

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -3083,7 +3083,7 @@ namespace ts {
         }
 
         function bindPropertyAssignment(name: BindableStaticNameExpression, propertyAccess: BindableStaticAccessExpression, isPrototypeProperty: boolean, containerIsClass: boolean) {
-            let namespaceSymbol = lookupSymbolForPropertyAccess(name);
+            let namespaceSymbol = lookupSymbolForPropertyAccess(name, container) || lookupSymbolForPropertyAccess(name, blockScopeContainer);
```
Comment (sandersn) [inline src/compiler/binder.ts:3086]: silly question: what happens if you *just* use blockScopeContainer? I assume that some other tests break.

File: tests/baselines/reference/topLevelBlockExpando.symbols [hunk 2]
```diff
@@ -0,0 +1,102 @@
+=== tests/cases/compiler/check.ts ===
+// 
+
+
+
+// https://github.com/microsoft/TypeScript/issues/31972
```
Comment (sandersn) [inline tests/baselines/reference/topLevelBlockExpando.symbols:6]: looks like tests are missing a baseline update

File: src/compiler/binder.ts (hunk 1 above)
Comment (andrewbranch) [reply src/compiler/binder.ts:3086]: I was pairing with Orta on this and I did try this, and stuff did fail. It was a little surprising to me, but I didn’t have time to dig into why that was the case.

Comment (weswigham) [reply src/compiler/binder.ts:3086]: `blockScopeContaininer` only contains the block scoped stuff. function-scoped stuff (like function declarations and `var`s), will only be in the function container (assuming they're different).

**Comment by orta** [comment]
This should be good for a second look

**Pull request merged**